pkg/review/bq/bq.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package bq contains abstracts away the boilerplate code that results from
+// mapping BigQuery table rows to and from a struct
+
+package bq
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"cloud.google.com/go/bigquery"
+	"google.golang.org/api/iterator"
+)
+
+// Client represents a bigquery client that automatically maps table
+// rows to and from structs of type T.
+type Client[T any] interface {
```

Comment:
What is the goal with this interface and what is the goal with making it generic? We seem to just be wrapping the native bigquery interface.

Reply:
I needed to have some indirection so I can test my business logic without actually making calls to BigQuery. 
See https://github.com/abcxyz/github-metrics-aggregator/blob/ce6de05391ff0b2d1a05905a987b710f2faf5dbc/pkg/review/commit_review_status_test.go#L1278
For how this is used in the tests.

As for making it generic, the client logic works regardless of struct type so I figured why not make it generic, especially since it seems like a common pattern to want ton take each row from a a BigQuery result and map it to a struct, so having a single piece of code to let you do that seemed like a useful thing.

Reply:
I think we might be conflating a few different problems in to the same solution. 

For testing, an interface is the correct answer. However, the interface should be _specific_ like `GetBreakGlassIssue`, not generic like `query`. For example:

```go
type BreakGlassIssuer interface {
  BreakGlassIssue(ctx context.Context, author, timestamp string) (*BreakglassIssue, error)
}
```

Then you can have multiple implementations of that interface:

```go
# real
type BigQueryBreakGlassIssuer struct {
  client *bigquery.Client
  config *Config
}

func NewBigQueryBreakGlassIssuer() (*BigQueryBreakGlassIssuer, error) {
  // TODO
}

type (b *BigQueryBreakGlassIssuer) BreakGlassIssue(ctx context.Context, author, timestamp string) (*BreakglassIssue, error) {
   // basically the same logic as L244 
}
```

```go
# test
type TestBreakGlassIssuer struct {
  results *BreakGlassIssue
  err error
}

type (b *BigQueryBreakGlassIssuer) BreakGlassIssue(ctx context.Context, author, timestamp string) (*BreakglassIssue, error) {
  return b.results, err
}
```

As for generics, the current implementation requires each _client_ to declare its structure up front; it does not implement a generic generic; you cannot have functions that return a different type that the struct. It seems really expensive to create a client for each data type I'd need to return.

Reply:
Yeah this is a limitation of Go's implementation of generics. Ideally you'd put the type parameter on the receiver function itself instead of the client, e.g. something like:
```
type Client interface {
	Query[T any](ctx context.Context, queryString string) ([]T, error)
}
```
but alas it's not possible to do that in Go.

I agree that that if you were working with many different types of structs having a Client each type would be expensive and clunky. But if you were mostly working with a one struct type (maybe your writing CRUD style API or something) then I don't think it would be that unreasonable. 

Personally, I don't think a generic interface has to be reasonable to use in every conceivable instance to justify its existence, just that there is a reusable bit of code that can be reasonably conceived to be reused in more than one instance (In this instance, mapping a big query table to a slice seemed like a common enough use case that making it generic seemed useful). 

You are correct though, for my testing needs all I need is to abstract is the the code to execute fetching the BreakGlassIssue. I'm not super married to this bit of abstraction so I've refactored it to be more along the lines of your suggestions.

Reply:
removed the client and refactored interface. let me know what you think.